feat(stepfunctions-tasks): add elasticmapreduce:AddTags permission for EmrCreateCluster state with tags

sjakthol opened the pull request:
Beginning April 24, 2023, Amazon Elastic Map Reduce (EMR) will start to require permission for the "elasticmapreduce:AddTags" IAM action when creating an EMR cluster with tags. This commit updates the EmrCreateCluster state to add elasticmapreduce:AddTags action to the state machine execution role when tags have been defined.

Closes #24842.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 52ed827492283aab9f395c4f9ee1b0f10b475283
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b19320e3-55aa-414e-93aa-cca3128f6bc0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 5dbf009dff4102e8702e4101926a417d49dfb180
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ae0e4f22-b8a6-4dd5-b3a0-da65b8633525%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/aws-stepfunctions-tasks/lib/emr/emr-create-cluster.ts:
```diff
@@ -319,6 +319,16 @@ export class EmrCreateCluster extends sfn.TaskStateBase {
+      policyStatements.push(
+        new iam.PolicyStatement({
+          actions: ['elasticmapreduce:AddTags'],
+          resources: ['*'],
```
Is it possible to scope this down further?

sjakthol replied:
As per https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonelasticmapreduce.html, AddTags action supports four kinds of EMR resources. Clusters are the only ones this state needs to be able to tag. The resource could be limited to all cluster resources (wildcard still required on the clusterid part of the ARN).

As for conditions, the action seems to support the usual RequestTag style condition keys. We could, in theory, add a condition to this statement that limits the tag keys and their values to the values specified in the tags prop. However, this gets complex pretty quickly since e.g. tag values can be dynamic jsonpath references to fields in the state object and tag keys can (maybe) even be CDK tokens. Constructing RequestTag conditions that support all these edge cases tricky very fast. Plus, if the user is allowed to use this CDK construct and create a role for the state machine, they could anyway specify any tags they wish in the props and hence the conditions generated based on them would allow any tags chosen by the user.  

A condition I would maybe like to see here is something that would allow tagging to be limited to happen during RunJobFlow call only. EC2 has ec2:CreateAction condition key which can be used to do that. With this kind of limitation you would not be able to go around tagging existing clusters if you happen to have access to execute a state machine that accepts tag keys / values from the state object. However, EMR does not seem to support such condition so that cannot be done.

Anyway, I'll change the resource to only allow tagging on clusters and leave out the additional conditions due to complications explained above. Thanks for the comment!

TheRealAmazonKendra reviewed (dismissed):
Thank you for your contribution! Please see my comments inline. Overall this looks great!

sjakthol commented on packages/@aws-cdk/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-with-tags.ts:
```diff
@@ -0,0 +1,29 @@
+  testCases: [stack],
+});
+
+app.synth();
```
Thanks. Removed in latest followup commit.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 99be1933cefd7bdcd8bb2cf42fc9a573e1cf9b82
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2c232b8b-bcfb-4bd5-9a14-207183409767%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

sjakthol commented:
Rebased on top of latest changes to main branch. Unfortunately I was not able to retain the commit history I had in this PR but had to recreate the changes in fresh commit on top of main branch instead. Comparing to previous state of this PR, the rebase included the following changes:
* Reapplied changes to aws-stepfunctions-tasks/lib/emr/emr-create-cluster.ts in the new location of this file
* Moved integration test to the new integration test package
No changes in content of the PR was made in this process.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: db7865274a85c580f1f7d9056bee8792a025dca4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=545a9f1c-04e1-4887-a79e-e5d5af513dbe%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: bec0bc530d72036767c88fd51cb35d8a3db1e0e9
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fddc6e4a-0b6e-4b4f-a28b-fded875fd1ee%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
